source-braintree-native/source_braintree_native/resources.py

```diff
@@ -0,0 +1,312 @@
+from datetime import timedelta
+import functools
+from logging import Logger
+
+from estuary_cdk.flow import CaptureBinding, ValidationError
+from estuary_cdk.capture import common, Task
+from estuary_cdk.http import HTTPMixin
+
+
+from .models import (
+    EndpointConfig,
+    ResourceConfig,
+    ResourceState,
+    FullRefreshResource,
+    FULL_REFRESH_RESOURCES,
+    Transaction,
+    Customer,
+    Dispute,
+    Subscription,
+    CreditCardVerification,
+)
+from .api import (
+    snapshot_resources,
+    fetch_transactions,
+    fetch_credit_card_verifications,
+    fetch_customers,
+    fetch_disputes,
+    fetch_subscriptions,
+)
+
+
+def _create_gateway(config: EndpointConfig) -> BraintreeGateway:
+    environment = braintree.Environment.Sandbox if config.advanced.is_sandbox else braintree.Environment.Production
+
+    return braintree.BraintreeGateway(
+        braintree.Configuration(
+            environment=environment,
+            merchant_id=config.merchant_id,
+            public_key=config.credentials.public_key,
+            private_key=config.credentials.private_key,
+        )
+    )
+
+def validate_credentials(
+        log: Logger, config: EndpointConfig
+):
+    gateway = _create_gateway(config)
+
+    try:
+        gateway.discount.all()
+    except AuthenticationError:
+        msg = f"Encountered issue while validating credentials. Please confirm provided endpoint configuration and API key credentials are correct."
+        raise ValidationError([msg])
+
+
+def full_refresh_resources(
+        log: Logger, config: EndpointConfig,
+) -> list[common.Resource]:
+
+    def open(
+            gateway: BraintreeGateway,
+            gateway_property: str,
+            gateway_response_field: str | None,
+            binding: CaptureBinding[ResourceConfig],
+            binding_index: int,
+            state: ResourceState,
+            task: Task,
+            all_bindings,
+    ):
+        common.open_binding(
+            binding,
+            binding_index,
+            state,
+            task,
+            fetch_snapshot=functools.partial(
+                snapshot_resources,
+                gateway,
+                gateway_property,
+                gateway_response_field,
+            ),
+            tombstone=FullRefreshResource(_meta=FullRefreshResource.Meta(op="d"), id=None)
+        )
+
+    return [
+        common.Resource(
+            name=name,
+            key=["/id"],
+            model=FullRefreshResource,
+            open=functools.partial(open, _create_gateway(config), gateway_property, gateway_response_field),
+            initial_state=ResourceState(),
+            initial_config=ResourceConfig(
+                name=name, interval=timedelta(minutes=5)
+            ),
+            schema_inference=True,
+        )
+        for (name, gateway_property, gateway_response_field) in FULL_REFRESH_RESOURCES
+    ]
+
+
+def transactions(
```

Comment:
Are all of these functions identical except for the `fetch_foobar` function and the resource name? Because it feels like this code could be deduplicated a lot if there's no reason to expect them to diverge in the near future.

Reply:
Yep, you're right. They're all the same except for the `fetch_foobar` functions and resource names. I'll refactor.